packages/devtools_app/lib/src/screens/debugger/program_explorer.dart

```diff
@@ -289,25 +291,57 @@ class _FileExplorer extends StatelessWidget {
   final Function(VMServiceObjectNode) onItemSelected;
   final Function(VMServiceObjectNode) onItemExpanded;
 
+  double get selectedNodeOffset =>
+      controller.selectedNodeIndex.value * _programExplorerRowHeight -
+      _selectedNodeTopSpacing;
+
+  @override
+  State<_FileExplorer> createState() => _FileExplorerState();
+}
+
+class _FileExplorerState extends State<_FileExplorer> with AutoDisposeMixin {
+  final ScrollController _scrollController = ScrollController();
+
+  @override
+  void initState() {
+    super.initState();
+    addAutoDisposeListener(
+        widget.controller.selectedNodeIndex, _maybeScrollToSelectedNode);
+  }
+
   @override
   Widget build(BuildContext context) {
-    return TreeView<VMServiceObjectNode>(
-      itemExtent: _programExplorerRowHeight,
-      dataRootsListenable: controller.rootObjectNodes,
-      onItemSelected: onItemSelected,
-      onItemExpanded: onItemExpanded,
-      dataDisplayProvider: (node, onTap) {
-        return _ProgramExplorerRow(
-          controller: controller,
-          node: node,
-          onTap: () {
-            controller.selectNode(node);
-            onTap();
-          },
-        );
-      },
+    WidgetsBinding.instance
+        .addPostFrameCallback((_) => _maybeScrollToSelectedNode());
+    return Scrollbar(
+      isAlwaysShown: true,
+      controller: _scrollController,
+      child: TreeView<VMServiceObjectNode>(
+        itemExtent: _programExplorerRowHeight,
+        dataRootsListenable: widget.controller.rootObjectNodes,
+        onItemSelected: widget.onItemSelected,
+        onItemExpanded: widget.onItemExpanded,
+        scrollController: _scrollController,
+        dataDisplayProvider: (node, onTap) {
+          return _ProgramExplorerRow(
+            controller: widget.controller,
+            node: node,
+            onTap: () {
+              widget.controller.selectNode(node);
+              onTap();
+            },
+          );
+        },
+      ),
     );
   }
+
+  void _maybeScrollToSelectedNode() {
+    if (_scrollController.offset != widget.selectedNodeOffset) {
```

Comment:
instead of checking for equality of offsets, can we check whether the selected node is already in view based on the offset and scroll extent? This will prevent spurious scrolls when the item is already visible.

Reply:
Good idea! Switched to that